src/RequestParser.cpp

```diff
@@ -817,3 +925,125 @@ bool RequestParser::isMethodAllowedToHaveBody(HTTPRequest& request)
+ * @param hasAlpha A reference to a boolean flag that will be set to true if the character is alphabetic.
+ * @return true if the character is a valid hostname character, false otherwise.
+ */
+bool RequestParser::isValidHostnameChar(char character, bool& hasAlpha)
+{
+	if (std::isalpha(character) != 0)
+		hasAlpha = true;
+	return ((std::isalnum(character) != 0) || character == '-');
+}
+
+/**
+ * @brief Validates a label within a hostname.
+ *
+ * This function checks if a given label within a hostname is valid according to the following rules:
+ * - The label must not be empty and must not exceed the maximum label length (RFC 1035).
+ * - The label must not start or end with a hyphen ('-').
+ * - All characters in the label must be valid hostname characters (alphanumeric or hyphen).
+ *
+ * @param label The label string to be validated.
+ * @return true if the label is valid, false otherwise.
+ */
+bool RequestParser::isValidLabel(const std::string& label, bool& hasAlpha)
+{
+	if (label.empty() || label.length() > s_maxLabelLength)
+		return false;
+
+	// Label must not start or end with a hyphen
+	if (label.at(0) == '-' || label.at(label.length() - 1) == '-')
+		return false;
+
+	for (size_t i = 0; i < label.length(); ++i)
+		if (!isValidHostnameChar(label.at(i), hasAlpha))
+			return false;
+	return true;
+}
+
+/**
+ * @brief Validates if the given hostname is valid according to specific rules.
+ *
+ * This function checks if the hostname length does not exceed the maximum allowed length
+ * and if each label (substring between periods) within the hostname is valid.
+ * Also verifies that the hostname has at least one alphabetic character.
+ *
+ * @param hostname The hostname to be validated.
+ * @return true if the hostname is valid, false otherwise.
+ */
+bool RequestParser::isValidHostname(const std::string& hostname)
+{
+	if (hostname.length() > s_maxHostNameLength)
+		return false;
+
+	size_t labelStart = 0;
+	size_t labelEnd = 0;
+	bool hasAlpha = false;
+
+	// Split hostname by periods and validate each label
+	while (labelEnd != std::string::npos) {
+		labelEnd = hostname.find('.', labelStart);
+
+		std::string label = (labelEnd == std::string::npos) ? hostname.substr(labelStart)
+															: hostname.substr(labelStart, labelEnd - labelStart);
+		if (!isValidLabel(label, hasAlpha))
+			return false;
+		labelStart = labelEnd + 1;
+	}
+	return hasAlpha;
+}
+
+/**
+ * @brief Checks if the given host value is a valid IP address.
+ *
+ * This function parses the provided host value to determine if it is a valid
+ * IPv4 address. It also checks if the IP address includes a port number.
+ *
+ * @param hostvalue The host value to be checked.
+ * @param hasPort A reference to a boolean that will be set to true if the IP address includes a port number, false otherwise.
+ * @return true if the host value is a valid IP address, false otherwise.
+ */
+bool RequestParser::isIPAddress(const std::string& hostvalue, bool& hasPort)
```

Comment:
This function does a whole lot, which is also covered by isIpAddressValid() and isPortValid() 
(it even says so in doxygen: "to determine if it is a valid IPv4 address").

I wouldn't even differentiate between IP and hostname in error message.

If you want to differentiate, I don't think you need this function. In validateHostHeader()
- check if you find a colon ":" --> if yes it has to be IP:port
- check if (ipAddress.find_first_not_of("0123456789.") != std::string::npos) --> has to be IP
- else has to be hostname

Reply:
Very good suggestion, as such I like it. But how do you handle these special cases:

InvalidHostnameLikeIP -> Host: 377.3.1.9999
InvalidHostnameLikeIP2 -> Host: 127.3.43.1.1

These look like ip addresses, but are not and should be handled by the isValidHostname() function. The isIpAddressValid() function does not recognize the error as it only checks of the first three segments. 

Idea: let the isIpAddressValid() handle these cases and update it with the code of the isIPAddress() function:

```
std::vector<std::string> segments;
	size_t segmentStart = 0;
	size_t segmentEnd = 0;

	while (segmentEnd != std::string::npos) {
		segmentEnd = hostvalue.find('.', segmentStart);
		std::string segment = (segmentEnd == std::string::npos)
			? hostvalue.substr(segmentStart)
			: hostvalue.substr(segmentStart, segmentEnd - segmentStart);
		segments.push_back(segment);
		segmentStart = segmentEnd + 1;
	}

if (segments.size() != 4)
		return false;
for (std::vector<std::string>::const_iterator citer = segments.begin(); citer != segments.end(); ++citer) {
		if (citer->empty() || citer->size() > 3)
			return false;
}
... then the existing code to check if is larger than 255 or lower than 0.
```

Reply:
I rewrote the isIPAddressValid() to incorporate the check for amount of segments and segment size:
```
bool isIpAddressValid(const std::string& ipAddress)
{
	if (ipAddress.find_first_not_of("0123456789.") != std::string::npos)
		return false;

	std::vector<std::string> segments;
	size_t segmentStart = 0;
	size_t segmentEnd = 0;

	while (segmentEnd != std::string::npos) {
		segmentEnd = ipAddress.find('.', segmentStart);
		std::string segment = (segmentEnd == std::string::npos)
			? ipAddress.substr(segmentStart)
			: ipAddress.substr(segmentStart, segmentEnd - segmentStart);
		segments.push_back(segment);
		segmentStart = segmentEnd + 1;
	}
	if (segments.size() != 4)
		return false;

	const short maxIpValue = 255;
	const short minIpValue = 0;
	const int base = 10;
	for (std::vector<std::string>::const_iterator citer = segments.begin(); citer != segments.end(); ++citer) {
		if (citer->empty() || citer->size() > 3)
			return false;
		const long segmentValue = std::strtol(citer->c_str(), NULL, base);
		if (segmentValue > maxIpValue || segmentValue < minIpValue)
			return false;
	}
	return true;
}
```

Reply:
I would use your split() here:

`const std::vector<std::string> segments = webutils::split(hostvalue, ".");`

Maybe we should also add some tests? If I understood correctly an IP in form 127.3.43.1.1 would have been accepted as correct?

Reply:
Ah yes, very good hint! Changed. There are two tests in test_parseHeader_Headers.cpp, namely the aforementioned:

InvalidHostnameLikeIP
InvalidHostnameLikeIP2